R/validate-policy.R

```diff
@@ -1,26 +1,35 @@
 #' Validate Policy File
 #' 
 #' Flattens and validates policy file
-#'  @param policy YAML policy file
-#'  @param gs Gradescope data
-#'  @importFrom purrr map discard
-#'  @export
+#' @param policy YAML policy file
+#' @param gs Gradescope data
+#' @importFrom purrr map discard
+#' @export
 validate_policy <- function(policy, gs){
   policy <- flatten_policy(policy)
-  # drop categories with unavailable assignments
-  categories <- map(policy$categories, "category") |> unlist()
-  # assignments is a vector that also includes category names
-  assignments <- c(get_assignments(gs), categories)
-  # this is because some categories have only nested categories as their assignments
-  # these categories should not be dropped
-  policy$categories <- map(policy$categories, function(cat){
-    cat$assignments <- cat$assignments[cat$assignments %in% assignments]
-    if (length(cat$assignments) == 0){
-      return (NULL)
-    }
-    return (cat)
-  }) |>
-    purrr::discard(is.null)
+  prev_length <- 0
+  current_length <- length(policy$categories)
+  #keep dropping until no more drops necessary
+  while (prev_length != current_length){
+    prev_length <- length(policy$categories)
+    categories <- map(policy$categories, "category") |> unlist()
+    assignments <- c(get_assignments(gs), categories)
+    policy$categories <- map(policy$categories, function(cat){
+      # drop categories with unavailable assignments/nested categories
+      cat$assignments <- cat$assignments[cat$assignments %in% assignments]
+      if (length(cat$assignments) == 0){
+        #if category has no assignments, dropp
+        return (NULL)
+      }
+      return (cat)
+    }) |>
+      purrr::discard(is.null)
```

Comment:
This looks like the step that bears on the questions asked in the thread: what is the behavior we want in situations where the policy file specifies stuff that's not available?

Reply:
The main accumulation point of this process is basically an extrapolation of this: in the first week of class, you only have say 5 IQs and nothing else, so even though RQs are only 3% of your final grade, they are currently 100% of your current grade. Even though Labs are in the syllabus, they are not currently relevant, and are hence dropped. I'm not entirely sure what setting Labs to NA (instead of dropping them) would do... I think it would cause trouble because if Lab 1 = NA, Lab 2 = NA, etc. Labs = mean(c(NA, NA), na.rm = TRUE) = NaN... so that's why we're dropping

Reply:
This is the understanding I had. However, it may not work exactly as intended right now. To build on the Nikita's example, if RQ's are only 3% of the grade, they would have weight 0.03 in the policy file. Hence, if all other categories are dropped, would not grades be capped at 3%? Perhaps we need to reweight weights by the sum of weights if we want this to work.

Reply:
Nice!

Reply:
The main reason that I could see to persist empty categories into the output is the data.table showing up in the dashboard. If I were a teacher, I could imagine it feeling natural to start the semester (if your policy file is complete) by looking at a data.table with all of the cols that I'd expect, but with no values filled in. As the semester progress, those empty cells get populated.